### PR TITLE
Initial edit of workflows documentation WIP - notes to be addressed

### PR DESCRIPTION
**Pull request by RHODOK**
Signed-off-by: Rho <Rho@Rho.com>

### CODE REVIEW

File: docs/source/index.rst [hunk 1]
```diff
@@ -5,33 +5,35 @@ Tao of workflows
...
+* **Primitives have clearly defined behavior.** It is up to the client to follow those definitions, otherwise it violates the purpose of being implementationally indepedent. ASK_MICKY (does this convey the intention?)
 
-* **Context is transparent looking up the stack but not down.** This means a subflow can see everything in its parent can see but the parent can not look into its subflow. A parent flow only "sees" the result of the subflow. Think of a class function and its return value.
+* **Context is transparent up the stack but not down.** This means a subflow can access everything its parent can, but the parent can not do likewise for a subflow. A parent flow only "sees" the result of the subflow. Think of a class function and its return value.
 
-* **Write tools.** Extending using tools allows for backwards compatibility while also making usage richer.
+* **Write tools.** Extension via tooling allows for backwards compatibility, while also making usage richer. ASK_MICKY (making for a richer usage experience? what do we mean by usage richer?)
```
Comment (RHODOK) [inline docs/source/index.rst]: "can do more with less"

File: docs/source/index.rst [hunk 2]
```diff
@@ -5,33 +5,35 @@ Tao of workflows
...
 
-* **Workflow primitives should be kept to a minimum.** It needs fit in someones head to be understandable. This is another place there tooling is key if you can construct a feature out of primitives do it. Then hide the construction behind tooling. This allows the things required to implement in the client to be minimal.
+* **Workflow primitives should be kept to a minimum.** It needs to fit in a developers head to be understandable and have utility. This is another place there tooling is key; if you can construct a feature from primitives, do. Hide this construction behind tooling. This allows the things requiring implementation in the client to be minimal.
 
-* **Primitives have clearly defined behavior.** It is up to the client to follow those definitions or it breaks the point of being able to drive multiple clients.
+* **Primitives have clearly defined behavior.** It is up to the client to follow those definitions, otherwise it violates the purpose of being implementationally indepedent. ASK_MICKY (does this convey the intention?)
```
Comment (RHODOK) [inline docs/source/index.rst]: reword this

File: docs/source/index.rst (hunk 1 above)
Comment (RHODOK) [reply docs/source/index.rst]: tools allow for complex composition, without the overhead of direct implementation - you can do more with less

File: docs/source/index.rst [hunk 3]
```diff
@@ -40,7 +42,7 @@ Glossary
 This is a glossary of terms used throughout this documentation.
 
 * **Context** is the datastore.
-* **Subflow** is a flow/loop task which is a subtask of another flow.
+* **Subflow** is a flow/loop task which is a subtask ASK_MICKY(is just a task of another flow right? dont think we need subtask, makes it sound like a different abstraction) of another flow.
```
Comment (RHODOK) [inline docs/source/index.rst]: make clear that a subtask is the same as a task, but exists on another task - this is true of all tasks that aren't the root flow

File: docs/source/primitives.rst [hunk 4]
```diff
@@ -8,25 +8,29 @@ Primitives
 jsonpath
 ########
 
-jsonpaths are used extensively throughout workflows for looking up objects. The full docs can be found `here <https://goessner.net/articles/JsonPath/>`_.
+jsonpaths are used extensively throughout workflows for object lookups. Full documentation can be found `here <https://goessner.net/articles/JsonPath/>`_.
 
-A quick rundown for the sake to expediency is that a jsonpath is a way of representing a look up in a structured object for instance given the object::
+A quick primer for the sake of expediency: a jsonpath a represention for a lookup in a structured object. For instance, given the object::
```
Comment (RHODOK) [inline docs/source/primitives.rst]: is a*

File: docs/source/primitives.rst [hunk 5]
```diff
@@ -40,21 +44,21 @@ Schema
...
 
 .. note:: The ``result`` object should copied then updated by parsing the ``result_keys`` this allows meta-data/debug-info to be set by the provider of the workflow.
 
-The `result` is either placed at ``destination_path`` or if ``destination_path = False`` the ``result`` is merged directly into the context above.
+The `result` is either placed at ``destination_path`` or if ``destination_path = False`` the ``result`` is merged directly into the context above ASK_MICKY (parent context?).
 
 
-.. note:: ``destination_path = False`` means merge result with the above context where as ``destination_path = None`` (the default value) means no destination path is set, this maybe used for a flow or loop with no ``result``
+.. note:: ``destination_path = False`` means merge result with the above context, whereas ``destination_path = None`` (the default value) means no destination path is set, this maybe used for a flow or loop which returns no ``result``
```
Comment (RHODOK) [inline docs/source/primitives.rst]: may be

Comment (RHODOK) [reply docs/source/primitives.rst]: does not define a result i.e. has no result keyword

File: docs/source/primitives.rst [hunk 6]
```diff
@@ -40,21 +44,21 @@ Schema
...
 
-.. warning:: Subflows (a flow within another flow) and see the context of the flow above it. However only state which is modified via the result objects will persist after leaving the flow.
+.. warning:: Subflows (a flow within another flow) can see the context of the flow above them. However, only state which is modified via the result objects will persist after leaving the flow.
 
 
-A flow can be just as list of tasks to be performed, a :ref:`while_loop_task` or a :ref:`for_loop_task`. Looping tasks build a list of `result` objects if one is defined.
+A flow can be as simple as a list of tasks to be performed, a :ref:`while_loop_task` or a :ref:`for_loop_task`. Looping tasks build a list of `result` objects if one is defined. ASK_MICKY (this seems awkward and disjointed, maybe we want to separate these somehow?)
```
Comment (RHODOK) [inline docs/source/primitives.rst]: These looping tasks

File: docs/source/primitives.rst [hunk 7]
```diff
@@ -64,20 +68,20 @@ A flow can be just as list of tasks to be performed, a :ref:`while_loop_task` or
 While loop
 ----------
 
-Repeat the flow tasks until a condition fails, the condition is a set of validators once one of these validators fails the loop is broken and the result is inserted into the context above.
-
+Repeat the flow tasks until a condition fails. The condition is a set of validators, when any of these validators fail the loop will continue its current iteration but will not restart, the loop is in effect broken and the result is inserted into the context above.
+ASK_MICKY (probably need to reword this, but just trying to make it clear that it doesn't break immediately, it follows the normal pattern for a while and doesn't restart - maybe we should mention the concept of a break event here to illustrate how to get out immediately)
```
Comment (RHODOK) [inline docs/source/primitives.rst]: note - validators are only checked at the start of a loop, so the loop will not enter the NEXT iteration

File: docs/source/primitives.rst [hunk 8]
```diff
@@ -64,20 +68,20 @@ A flow can be just as list of tasks to be performed, a :ref:`while_loop_task` or
...
 
 .. _for_loop_task:
 
 For loop
 --------
 
-Repeat the flow tasks for a given list of objects. Each iteration the object at that index of the list is merged into the context and then the tasks are evaluated. Once the list of objects has been exhausted the loop will break and the result will be inserted into the context above. For example assuming the ``iterable_path="$.for_loop_list"`` and the context is before:
+Repeat the flow tasks for a given list of objects. At each iteration the object at the current index of the list is merged into the context, the flows tasks are then evaluated. Once the list of objects has been exhausted, the loop will exit at the end of its current iteration and the result will be inserted into the context above. For example assuming the ``iterable_path="$.for_loop_list"`` and the context bagins as:
```
Comment (RHODOK) [inline docs/source/primitives.rst]: "this flows"

File: docs/source/primitives.rst [hunk 9]
```diff
@@ -126,7 +130,7 @@ Schema
 Screen
 ******
 
-Screens are the only task type which display components to the screen (excluding status messages which can be presented by other tasks although they will be shown on the next screen task presented to the user).
+Screen are the sole task type used to display components to the user (excluding status messages which can be presented by other tasks, although they will be shown on the next screen task presented to the user).
```
Comment (RHODOK) [inline docs/source/primitives.rst]: Screen tasks?

File: docs/source/primitives.rst [hunk 10]
```diff
@@ -152,27 +156,27 @@ Schema
 Update
 ******
 
-Update tasks are used to change values in the context. The subtasks have three major forms:
+Update tasks are used to change values in the context. These tasks have three major forms:
```
Comment (RHODOK) [inline docs/source/primitives.rst]: this needs fleshing out to capture the idea that an update task itself contains multiple tasks, of the three major forms - the thing they do is defined by the structure of the task (what keyword - maybe list this in someway?)

File: docs/source/primitives.rst [hunk 11]
```diff
@@ -152,27 +156,27 @@ Schema
...
 
-3. creating a new value using a string template
+3. creating a new value using a string template:
 
 .. code-block::
 
     {"template": "{$.value}{$.another}", "result_key": "$.destination"}
 
-There are also extra flags ``append`` and ``extend`` which are mutually exclusive which act by updating the destination which is assumed to be a list.
+There are also extra flags ``append`` and ``extend``, which are mutually exclusive, they act by updating the destination (assumed to be a list) ASK_MICKY (are they always a list? This makes it sound too optional, we should state about cases where they aren't if they aren't).
```
Comment (RHODOK) [inline docs/source/primitives.rst]: they are a list, anything else will fail

File: docs/source/primitives.rst [hunk 12]
```diff
@@ -272,7 +276,7 @@ Schema
 Components
 ##########
 
-Components are screen elements to be displayed to and interacted by the user. Components are split into two parts. The base component and the component look up. The base component is extracted into :ref:`components key <basic_structure>` in the workflow which is then used by the component look up, because of this components with the same name are required to have the same values, otherwise an error is thrown. As with tasks components have preconditions which dictate if they are shown.
+Components are screen elements to be displayed to, and interacted with by, the user. Components are split into two parts. The base component and the component look up. The base component is extracted into :ref:`components key <basic_structure>` in the workflow, this is then used by the component look up, as such components with the same name are required to have the same values, otherwise an error is thrown. As with task components, these have preconditions which dictate if they are shown.
```
Comment (RHODOK) [inline docs/source/primitives.rst]: try and rephrase this starting sentence

File: docs/source/primitives.rst (hunk 8 above)
Comment (RHODOK) [inline docs/source/primitives.rst]: filthy bagins'...

Comment (RHODOK) [reply docs/source/primitives.rst]: eep

File: docs/source/primitives.rst [hunk 13]
```diff
@@ -50,15 +50,15 @@ Returning values from the context scope to the flow above is done by setting the
 
 .. note:: The ``result`` object should copied then updated by parsing the ``result_keys`` this allows meta-data/debug-info to be set by the provider of the workflow.
 
-The `result` is either placed at ``destination_path`` or if ``destination_path = False`` the ``result`` is merged directly into the context above ASK_MICKY (parent context?).
+The `result` is either placed at ``destination_path`` or if ``destination_path = False`` the ``result`` is merged directly into the parent context, that is the context of the flow/task of which this is a subordinate. ASK_MICKY (have i made it better or worse?)
```
Comment (RHODOK) [inline docs/source/primitives.rst]: only flow tasks have tasks - redraft with that in mind. Subordinate is confusing, subtask as used elsewhere probably

File: docs/source/primitives.rst [hunk 14]
```diff
@@ -68,15 +68,13 @@ A flow can be as simple as a list of tasks to be performed, a :ref:`while_loop_t
 While loop
 ----------
 
-Repeat the flow tasks until a condition fails. The condition is a set of validators, when any of these validators fail the loop will continue its current iteration but will not restart, the loop is in effect broken and the result is inserted into the context above.
-ASK_MICKY (probably need to reword this, but just trying to make it clear that it doesn't break immediately, it follows the normal pattern for a while and doesn't restart - maybe we should mention the concept of a break event here to illustrate how to get out immediately)
-
+Repeat the flow tasks until a condition fails. The condition is a set of validators, when any of these validators fail the loop will exit at the end of its current iteration, the loop is in effect broken and the result is inserted into the context above. In order to leave the loop partway through an iteration a break event can be used ASK_MICKY(INSERT_BREAK_EVENT_LINK)
```
Comment (RHODOK) [inline docs/source/primitives.rst]: event with action break

File: docs/source/primitives.rst [hunk 15]
```diff
@@ -156,27 +154,27 @@ Schema
...
 
-1. updating the context with a raw value:
+1. updating the context with a raw value, key word - "result":
 
 .. code-block::
 
     {"result": "x", "result_key": "$.destination"}
 
-2. renaming a context value:
+2. renaming a context value, key word - "key":
```
Comment (RHODOK) [inline docs/source/primitives.rst]: keyword

Comment (RHODOK) [reply docs/source/primitives.rst]: bracket it

File: docs/source/primitives.rst [hunk 16]
```diff
@@ -156,27 +154,27 @@ Schema
...
 
-3. creating a new value using a string template:
+3. creating a new value using a string template - key word - "template":
 
 .. code-block::
 
     {"template": "{$.value}{$.another}", "result_key": "$.destination"}
 
-There are also extra flags ``append`` and ``extend``, which are mutually exclusive, they act by updating the destination (assumed to be a list) ASK_MICKY (are they always a list? This makes it sound too optional, we should state about cases where they aren't if they aren't).
+There are also extra flags ``append`` and ``extend``, which are mutually exclusive, they act by modifying the above major cases ASK_MICKY (is this correct?), updating a destination which must be a list.
```
Comment (RHODOK) [inline docs/source/primitives.rst]: remove the ask

Comment (RHODOK) [reply docs/source/primitives.rst]: extend requires the result you are trying to add on to also be a list, the difference between an append and an extend

File: docs/source/extending.rst [hunk 17]
```diff
@@ -1,21 +1,21 @@
 Extending workflows
 ###################
 
-Workflows is meant to be extensible this is mainly done defining new class either by subclassing. Things which may require you to extend workflows are things like a new type of action you wish to preform like some form which is not currently supported like evaluating a piece of JS.
+Workflows is meant to be extensible, this is typically achieved by defining a new class from scratch or from an existing class via inheritance, through use of a subclass. Needs necessitating workflows extension are those requiring a new type of action you wish to perform, one which is not currently supported, such evaluating a novel piece of JS. ASK_MICKY (js function?)
```
Comment (RHODOK) [inline docs/source/extending.rst]: such as

Comment (RHODOK) [reply docs/source/extending.rst]: needs necessitating? did you fall into a vat of redundancy

File: docs/source/extending.rst [hunk 18]
```diff
@@ -1,21 +1,21 @@
...
 
 
 
 Adding new tasks types
 **********************
 
-Adding tasks to the workflows_engine is simple the complicated bit will always be the parsing and execution of the task.
-The only thing to wary of is that you must extend ``TASK_TYPE_MAPPING`` within ``workflows_engine.core.tasks`` otherwise ``Flow.add_task`` will not pick up you new task. This is also how you could overwrite behavior of a predefined task.
+Adding tasks to the workflows_engine is simple, the complicating aspect will always be the parsing and execution of the task.
+The primary concern of which to be wary is that you must extend ``TASK_TYPE_MAPPING`` within ``workflows_engine.core.tasks`` otherwise ``Flow.add_task`` will not pick up the new task. This is the approach to use in order to ASK_MICKY (when you said could here, do you mean this is how you would want to achieve that?) overwrite behavior of a predefined task.
```
Comment (RHODOK) [inline docs/source/extending.rst]: this is how you have to do it, but you may not want to

File: docs/source/primitives.rst [hunk 19]
```diff
@@ -152,27 +154,27 @@ Schema
...
 
-3. creating a new value using a string template
+3. creating a new value using a string template (key word - "template"):
 
 .. code-block::
 
     {"template": "{$.value}{$.another}", "result_key": "$.destination"}
 
-There are also extra flags ``append`` and ``extend`` which are mutually exclusive which act by updating the destination which is assumed to be a list.
+There are also extra flags ``append`` and ``extend``, which are mutually exclusive, they act by modifying the above major cases, updating a destination which must be a list. Extend and append are used with the conventional meaning, such that extend will add a list and append will a value, as such the result for extend and append are a list and a value respectively. ASK_MICKY (maybe I should make the sentence a bit more colloquial though bad, "the result for extend is a list and the result for append is a value")
```
Comment (RHODOK) [inline docs/source/primitives.rst]: will add a value - use concatenate instead

File: docs/source/primitives.rst [hunk 20]
```diff
@@ -199,7 +201,7 @@ Schema
 Condition
 *********
 
-Selects (jumps to) a task to switch to based on if a condition is true or false.
+Selects (jumps to) a task to switch to based on if a condition is true or false. ASK_MICKY [mick, didn't you say you had an update for this?](we should specify that this is only one in a flow we have been in and is defined on its flow I think)
 
 .. warning:: Jumps are only allowed to the same flow or the flows parents.
```
Comment (unirho) [inline docs/source/primitives.rst:206]: flows parents - using the flow name is also

File: docs/source/index.rst [hunk 21]
```diff
@@ -5,33 +5,35 @@ Tao of workflows
...
 
-* **Minimize surprise.** We should try to make things as least surprising as possible if there is a direct comparison with something in a programing language it should act the same.
+* **Minimize surprise.** We should try to make things the least surprising as possible: if there is a direct comparison with something in a programing language it should act the same.
 
-* **Workflow primitives should be kept to a minimum.** It needs fit in someones head to be understandable. This is another place there tooling is key if you can construct a feature out of primitives do it. Then hide the construction behind tooling. This allows the things required to implement in the client to be minimal.
+* **Workflow primitives should be kept to a minimum.** It needs to fit in a developers head to be understandable and have utility. This is another place there tooling is key; if you can construct a feature from primitives, do. Hide this construction behind tooling. This allows the things requiring implementation in the client to be minimal.
```
Comment (RHODOK) [inline docs/source/index.rst]: hmmmmmmmmmmmmmmmmmmmmmmmm the last sentence still doesn't read right

Comment (RHODOK) [reply docs/source/index.rst]: This allows the things the client is required to implement in the client to be minimal

File: docs/source/primitives.rst [hunk 22]
```diff
@@ -199,7 +201,7 @@ Schema
 Condition
 *********
 
-Selects (jumps to) a task to switch to based on if a condition is true or false.
+Selects (jumps to) a task to switch to based on if a condition is true or false. ASK_MICKY [mick, didn't you say you had an update for this?](we should specify that this is only one in a flow we have been in and is defined on its flow I think)
```
Comment (RHODOK) [inline docs/source/primitives.rst]: flows parents - using the flow name is also

File: docs/source/quickstart.rst [hunk 23]
```diff
@@ -6,7 +6,8 @@ Quick start
     import json
     from workflows_engine import Workflow
 
-Here is a simple example of a workflow. It shows a screen with an input and a submit button. The input field has a validation which is shown if the length of the value in the input is zero. When the button  is clicked either the validators error message will be shown below the input or another screen will be shown with a message info box containing the string input into the field.
+Below can be found a simple example of a workflow implementation and the resultant workflow json. It is simply a screen with an input and a submit button. The input field itself is subject to a validtor, this checks that the length of an input given is greater than 0. When the button is clicked either the validator error message will be shown below the input (should the length be 0 i.e. an empty input) or the user will progress to another screen, which will present an info box containing the string input into the field.
```
Comment (RHODOK) [inline docs/source/quickstart.rst]: validator

File: docs/source/quickstart.rst [hunk 24]
```diff
@@ -6,7 +6,8 @@ Quick start
...
+This acts to demonstrate several concepts described previously such as the validation of an input and transition between screen components, as well as more technical curiosities such as implementing a new validator. This is not intended to serve as an exhaustive representation of workflows, but simply a gentle introduction into how it can be put together.
```
Comment (RHODOK) [inline docs/source/quickstart.rst]: described elsewhere

File: docs/source/extending.rst [hunk 25]
```diff
@@ -1,7 +1,7 @@
 Extending workflows
 ###################
 
-Workflows is meant to be extensible, this is typically achieved by defining a new class from scratch or from an existing class via inheritance, through use of a subclass. Needs necessitating workflows extension are those requiring a new type of action you wish to perform, one which is not currently supported, such evaluating a novel piece of JS. ASK_MICKY (js function?)
+Workflows is meant to be extensible, this is typically achieved by defining a new class from scratch or from an existing class via inheritance, through use of a subclass. The need for workflows extension arises from requiring a new type of action you wish to perform, one which is not currently supported, such evaluating a novel piece of JS.
```
Comment (RHODOK) [inline docs/source/extending.rst]: such as

File: docs/source/primitives.rst [hunk 26]
```diff
@@ -68,7 +68,7 @@ A flow can be as simple as a list of tasks to be performed, a :ref:`while_loop_t
 While loop
 ----------
 
-Repeat the flow tasks until a condition fails. The condition is a set of validators, when any of these validators fail the loop will exit at the end of its current iteration, the loop is in effect broken and the result is inserted into the context above. In order to leave the loop partway through an iteration a break event can be used ASK_MICKY(INSERT_BREAK_EVENT_LINK)
+Repeat the flow tasks until a condition fails. The condition is a set of validators, when any of these validators fail the loop will exit at the end of its current iteration, the loop is in effect broken and the result is inserted into the context above. In order to leave the loop partway through an iteration an event with an action of type break can be used ASK_MICKY(INSERT_BREAK_EVENT_LINK)
```
Comment (RHODOK) [inline docs/source/primitives.rst]: :ref:`Event <event_task>`

File: docs/source/primitives.rst [hunk 27]
```diff
@@ -8,25 +8,29 @@ Primitives
...
 Tasks
 #####
 
-Tasks are actions, to be preformed by the client and take the form of specific objects types which is identified by the ``type`` keyword.
+Tasks are actions to be performed by the client and take the form of specific objects types identified by the ``type`` keyword.
 
 .. important::
 
-    * All task names should be unique within a flow.
+    * All task names should be unique within a flow (remember, subflows are themselves a flow ASK_MICKY (INSERT LINK TO INDEX DEFINITION)).
```
Comment (RHODOK) [inline docs/source/primitives.rst]: .. _glossary: - in the index

:ref:`subflows <glossary>`

File: docs/source/structure.rst [hunk 28]
```diff
@@ -13,5 +13,8 @@ The ``validator`` key contains a list of :ref:`validator <validator_objects>` wh
 
 The ``flow`` key contains the base task for the workflow (think the ``main`` function in a Java program).
 
+The ``hash`` key is a hash for representing the workflow (this is intended for use by the client to check for changes to the workflow - no need to grab the definition from the server if you already have it)
+
+The ``context`` key contains the initial values used by the "Context" as defined in index ASK_MICKY (link to index?)
```
Comment (RHODOK) [inline docs/source/structure.rst]: .. _glossary: - in the index

:ref:subflows <glossary>

**Pull request merged**